druid/src/widget/viewport.rs

```diff
@@ -0,0 +1,259 @@
+// Copyright 2020 The Druid Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::kurbo::{Affine, Point, Rect, Size, Vec2};
+use crate::widget::prelude::*;
+use crate::{Data, WidgetPod};
+
+/// Represents the size and position of a rectangular "viewport" into a larger area.
+#[derive(Clone, Copy, Default, Debug, PartialEq)]
+pub struct ViewportPos {
+    /// The size of the area that we have a viewport into.
+    pub content_size: Size,
+    /// The view rectangle. This should be contained in `content_size.to_rect()`.
+    pub rect: Rect,
```

Comment:
What about the case of a scroll + zoom view that lets you decrease the zoom level so that the content is smaller than the allowed size? This is actually a case I encounter in runebender, and is a common pattern in things like image viewers?

It may be that I can handle that at another level of the stack, but wanted to raise the question; it's a case I'd like to support.

Reply:
Good point. I think it definitely makes sense for `ViewportPos` to support this. For `Viewport`, it probably raises some layout issues that I'd like to punt on for now.

Reply:
This seems to be mostly a documentation issue, since I'm neither enforcing nor relying on the view being smaller than the content. The panning methods are intended for that use case, so I've tried to make that clear in their docstrings.